File: sample/src/main/swift/Sources/SampleAppCore/Int32Test.swift
//
// Created by Andrew on 2/4/18.
//

import Foundation

public enum Int32Enum: Int32, Codable {
    case one
    case two
    case three
}

public struct Int32OptionsSet: OptionSet, Codable {

    public let rawValue: Int32

    public init(rawValue: Int32) {
        self.rawValue = rawValue
    }

    static let one = Int32OptionsSet(rawValue: 1 << 0)
    static let two = Int32OptionsSet(rawValue: 1 << 1)
    static let three = Int32OptionsSet(rawValue: 1 << 2)
}

public struct Int32TestStruct: Codable, Hashable {
    public var zero: Int32 = Int32.zero
    public var max: Int32 = Int32.max
    public var min: Int32 = Int32.min
    public var optional: Int32? = Int32.zero
    public var optionalNil: Int32? = nil
}

public protocol Int32TestParamProtocol {
    func testParam(_ param: Int32) -> Bool
}

public protocol Int32TestReturnTypeProtocol {
    func testReturnType() -> Int32
}

public protocol Int32TestOptionalParamProtocol {
    func testOptionalParam(_ param: Int32?) -> Bool
}

public protocol Int32TestOptionalReturnTypeProtocol {
    func testOptionalReturnType() -> Int32?
}

public typealias Int32Block = (_ value: Int32) -> Int32
public typealias OptionalInt32Block = (_ value: Int32?) -> Int32?

public class Int32Test {

    public static func testZero() -> Int32 {
        return 0
    }

    public static func testMin() -> Int32 {
        return Int32.min
    }

    public static func testMax() -> Int32 {
        return Int32.max
    }

    public static func testParam(_ param: Int32) -> Bool {
        return param == Int32.max
    }

    public static func testReturnType() -> Int32 {
        return Int32.max
    }

    public static func testOptionalParam(_ param: Int32?) -> Bool {
        return param == Int32.max
    }

    public static func testOptionalReturnType() -> Int32? {
        return Int32.max
    }

    public static func testProtocolParam(_ callback: Int32TestParamProtocol) -> Bool {
        return callback.testParam(Int32.max)
    }

    public static func testProtocolReturnType(_ callback: Int32TestReturnTypeProtocol) -> Int32 {
        return callback.testReturnType()
    }

    public static func testProtocolOptionalParam(_ callback: Int32TestOptionalParamProtocol) -> Bool {
        return callback.testOptionalParam(Int32.max)
    }

    public static func testProtocolOptionalReturnType(_ callback: Int32TestOptionalReturnTypeProtocol) -> Int32? {
        return callback.testOptionalReturnType()
    }

    public static func testEncode() ->  Int32TestStruct {
        return Int32TestStruct()
    }

    public static func testDecode(_ value: Int32TestStruct) -> Bool {
        return value == Int32TestStruct()
    }

    public static func testEnumEncode(_ rawValue: Int) -> Int32Enum {
        switch Int32(rawValue) {
        case Int32Enum.one.rawValue: return Int32Enum.one
        case Int32Enum.two.rawValue: return Int32Enum.two
        case Int32Enum.three.rawValue: return Int32Enum.three
        default: fatalError("Can't find enum with rawValue \(rawValue)")
        }
    }

    public static func testEnumDecode(_ enumValue: Int32Enum) -> Int {
        return Int(enumValue.rawValue)
    }

    public static func testOptionSetEncode(_ rawValue: Int) -> Int32OptionsSet {
        return Int32OptionsSet(rawValue: Int32(rawValue))
    }

    public static func testOptionSetDecode(_ optionSet: Int32OptionsSet) -> Int {
        return Int(optionSet.rawValue)
    }

    public static func testBlock(_ block: Int32Block) -> Bool {
        let value = block(0)
        return value == 0
    }

    public static func testOptionalBlock(_ block: OptionalInt32Block) -> Bool {
        let value = block(nil)
        return value == nil
    }

}

File: sample/src/main/java/com/readdle/swiftjava/sample/Int32Test.kt
package com.readdle.swiftjava.sample

import com.readdle.codegen.anotation.SwiftReference
import com.readdle.codegen.anotation.SwiftValue
import java.lang.annotation.Native

@SwiftValue
enum class Int32Enum(val rawValue: Int) {

    ONE(0), TWO(1), THREE(2);

    companion object {

        private val values = HashMap<Int, Int32Enum>()

        @JvmStatic
        fun valueOf(rawValue: Int): Int32Enum {
            return values[rawValue]!!
        }

        init {
            enumValues<Int32Enum>().forEach {
                values[it.rawValue] = it
            }
        }
    }

}

@SwiftValue
data class Int32OptionsSet(val rawValue: Int = 0) {
    companion object {
        @JvmStatic
        val  one = Int32OptionsSet(1)
        @JvmStatic
        val  two = Int32OptionsSet(2)
        @JvmStatic
        val  three = Int32OptionsSet(4)
    }
}

@SwiftValue
data class Int32TestStruct(var zero: Int = 0,
                           var max: Int = Int.MAX_VALUE,
                           var min: Int = Int.MIN_VALUE,
                           var optional: Int? = 0,
                           var optionalNil: Int? = null)

@SwiftReference
class Int32Test private constructor() {

    companion object {

        @JvmStatic
        external fun testEncode(): Int32TestStruct

        @JvmStatic
        external fun testDecode(value: Int32TestStruct): Boolean
        
        @JvmStatic
        external fun testEnumEncode(rawValue: Int) : Int32Enum

        @JvmStatic
        external fun testEnumDecode(enum: Int32Enum) : Int

        @JvmStatic
        external fun testOptionSetEncode(rawValue: Int) : Int32OptionsSet

        @JvmStatic
        external fun testOptionSetDecode(enum: Int32OptionsSet) : Int
    }

    @Native
    var nativePointer: Long = 0

    external fun release()

}